rueidisprob/countingbloomfilter.go

```diff
@@ -0,0 +1,411 @@
+		client:              client,
+		name:                bfName,
+		counter:             counterName,
+		hashIterations:      hashIterations,
+		hashIterationString: strconv.FormatUint(uint64(hashIterations), 10),
+		size:                size,
+		addMultiScript:      rueidis.NewLuaScript(countingBloomFilterAddMultiScript),
+		addMultiKeys:        []string{bfName, counterName},
+		removeMultiScript:   rueidis.NewLuaScript(countingBloomFilterRemoveMultiScript),
+		removeMultiKeys:     []string{bfName, counterName},
+	}, nil
+}
+
+func (f *countingBloomFilter) Add(ctx context.Context, key string) error {
+	return f.AddMulti(ctx, []string{key})
+}
+
+func (f *countingBloomFilter) AddMulti(ctx context.Context, keys []string) error {
+	if len(keys) == 0 {
+		return nil
+	}
+
+	indexes := f.indexes(keys)
+
+	args := make([]string, 0, len(indexes)+1)
+	args = append(args, strconv.Itoa(len(keys)))
+	args = append(args, indexes...)
+
+	resp := f.addMultiScript.Exec(ctx, f.client, f.addMultiKeys, args)
+	return resp.Error()
+}
+
+func (f *countingBloomFilter) indexes(keys []string) []string {
+	allIndexes := make([]string, 0, len(keys)*int(f.hashIterations))
+	size := uint64(f.size)
+	for _, key := range keys {
+		h1, h2 := hash([]byte(key))
+		for i := uint(0); i < f.hashIterations; i++ {
+			allIndexes = append(allIndexes, strconv.FormatUint(index(h1, h2, i, size), 10))
+		}
+	}
+	return allIndexes
+}
+
+func (f *countingBloomFilter) Exists(ctx context.Context, key string) (bool, error) {
+	exists, err := f.ExistsMulti(ctx, []string{key})
+	if err != nil {
+		return false, err
+	}
+
+	return exists[0], nil
+}
+
+func (f *countingBloomFilter) ExistsMulti(ctx context.Context, keys []string) ([]bool, error) {
+	if len(keys) == 0 {
+		return nil, nil
+	}
+
+	indexes := f.indexes(keys)
+
+	resp := f.client.Do(
+		ctx,
+		f.client.B().
+			Hmget().
+			Key(f.name).
+			Field(indexes...).
+			Build(),
+	)
+	if resp.Error() != nil {
+		return nil, resp.Error()
+	}
+
+	messages, err := resp.ToArray()
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]bool, 0, len(keys))
+	isExist := true
+	for i, message := range messages {
+		cnt, err := message.AsUint64()
+		if err != nil {
+			if !rueidis.IsRedisNil(err) {
+				return nil, err
+			}
+
+			isExist = false
+		}
+
+		if cnt == 0 {
+			isExist = false
+		}
+
+		if (i+1)%int(f.hashIterations) == 0 {
+			result = append(result, isExist)
+			isExist = true
+		}
+	}
+
+	return result, nil
+}
+
+func (f *countingBloomFilter) Remove(ctx context.Context, key string) error {
+	return f.RemoveMulti(ctx, []string{key})
+}
+
+func (f *countingBloomFilter) RemoveMulti(ctx context.Context, keys []string) error {
+	if len(keys) == 0 {
+		return nil
+	}
+
+	deduplicatedKeys := make([]string, 0, len(keys))
+	keySet := make(map[string]struct{}, len(keys))
+	for _, key := range keys {
+		if _, ok := keySet[key]; !ok {
+			keySet[key] = struct{}{}
+			deduplicatedKeys = append(deduplicatedKeys, key)
```

Comment:
Should we do this deduplication? It feels to me that duplication should be allowed.

Reply:
[f959dc3](https://github.com/redis/rueidis/pull/519/commits/f959dc3203e4538e83bbb5d8c35e410a2d720b3e)

To mitigate complex logic, i added deduplication. But I remove it it is more complex and can downgrade performance.

Reply:
I see. It is indeed complex but looks good to me now.